ArmoniK.Extension.CSharp.Client/Services/BlobService.cs

```diff
@@ -371,4 +404,181 @@ await blobClient.UploadResultData(blob.SessionId,
+        if (blobDefinition.TotalSize >= serviceConfiguration_.DataChunkMaxSize)
+        {
+          // For a blob size above the threshold of DataChunkMaxSize, we add it also to the list of blobs without data
+          // so that its metadata will be created by CreateBlobsMetadataAsync(). Subsequently, the upload will be processed by CreateBlobsWithContentAsync()
+          blobsWithoutData.Add(blobDefinition);
+        }
+      }
+      else
+      {
+        blobsWithoutData.Add(blobDefinition);
+      }
+    }
+
+    if (blobsWithoutData.Any())
+    {
+      // Creation of blobs without data
+      var blobsCreate = blobsWithoutData.Select(b => (b.Name, b.ManualDeletion));
+      var response = CreateBlobsMetadataAsync(session,
+                                              blobsCreate,
+                                              cancellationToken);
+      var index = 0;
+      await foreach (var blob in response.ConfigureAwait(false))
+      {
+        blobsWithoutData[index].BlobHandle = new BlobHandle(blob,
+                                                            armoniKClient_);
+        index++;
+      }
+    }
+
+    if (blobsWithData.Any())
+    {
+      // Creation of blobs with data
+      var response = CreateBlobsWithContentAsync(session,
+                                                 blobsWithData,
+                                                 cancellationToken);
+      var index = 0;
+      await foreach (var blob in response.ConfigureAwait(false))
+      {
+        blobsWithData[index].BlobHandle = new BlobHandle(blob,
+                                                         armoniKClient_);
+        index++;
+      }
+    }
+  }
+
+  private async IAsyncEnumerable<BlobInfo> CreateBlobsWithContentAsync(SessionInfo                                session,
+                                                                       IEnumerable<BlobDefinition>                blobDefinitions,
+                                                                       [EnumeratorCancellation] CancellationToken cancellationToken = default)
+  {
+    // This is a bin packing kind of problem for which we apply the first-fit-decreasing strategy
+    var batches = GetOptimizedBatches(blobDefinitions,
+                                      serviceConfiguration_.DataChunkMaxSize);
+
+    await using var channel = await channelPool_.GetAsync(cancellationToken)
+                                                .ConfigureAwait(false);
+    var blobClient = new Results.ResultsClient(channel);
+
+    foreach (var batch in batches)
+    {
+      if (serviceConfiguration_ != null && batch.Size > serviceConfiguration_.DataChunkMaxSize)
+      {
+        // We are in a case where the size is above the threshold, then the batch contains a single element
+        var blobDefinition = batch.Items.Single();
+        await UploadBlobByChunkAsync(blobClient,
+                                     blobDefinition)
+          .ConfigureAwait(false);
```

Comment:
Channel pool should record the exception
```suggestion
        try {
          await UploadBlobByChunkAsync(blobClient,
                                       blobDefinition)
            .ConfigureAwait(false);
        }
        catch (Exception ex)
        {
          channel.Exception = ex;
          throw;
        }
```